flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBNativeMetricMonitor.java

```diff
@@ -0,0 +1,154 @@
+import org.apache.flink.metrics.Gauge;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.metrics.View;
+import org.apache.flink.util.FlinkRuntimeException;
+import org.apache.flink.util.IOUtils;
+import org.apache.flink.util.ResourceGuard;
+
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+/**
+ * A monitor which pull {{@link RocksDB}} native metrics
+ * and forwards them to Flink's metric group. All metrics are
+ * unsigned longs and are reported at the column family level.
+ */
+@Internal
+public class RocksDBNativeMetricMonitor implements Closeable {
+
+	private final CloseableRegistry registeredGauges;
+
+	private final RocksDB db;
+
+	private final ResourceGuard.Lease lease;
+
+	private final RocksDBNativeMetricOptions options;
+
+	private final MetricGroup metricGroup;
+
+	RocksDBNativeMetricMonitor(
+		@Nonnull RocksDB db,
+		@Nonnull ResourceGuard guard,
+		@Nonnull RocksDBNativeMetricOptions options,
+		@Nonnull MetricGroup metricGroup
+	) throws IOException {
+		this.db = db;
+		this.lease = guard.acquireResource();
+		this.options = options;
+		this.metricGroup = metricGroup;
+
+		this.registeredGauges = new CloseableRegistry();
+	}
+
+	/**
+	 * Register gauges to pull native metrics for the column family.
+	 * @param columnFamilyName group name for the new gauges
+	 * @param handle native handle to the column family
+	 */
+	void registerColumnFamily(String columnFamilyName, ColumnFamilyHandle handle) {
+		try {
+			MetricGroup group = metricGroup.addGroup(columnFamilyName);
+
+			for (String property : options.getProperties()) {
+				RocksDBNativeMetricView gauge = new RocksDBNativeMetricView(
+					property,
+					handle,
+					db
+				);
+
+				group.gauge(property, gauge);
+				registeredGauges.registerCloseable(gauge);
+			}
+		} catch (IOException e) {
+			throw new FlinkRuntimeException("Unable to register native metrics with RocksDB", e);
+		}
+	}
+
+	@Override
+	public void close() {
+		IOUtils.closeQuietly(registeredGauges);
+		IOUtils.closeQuietly(lease);
+	}
+
+	static class RocksDBNativeMetricView implements Gauge<String>, View, Closeable {
+		private static final Logger LOG = LoggerFactory.getLogger(RocksDBNativeMetricView.class);
+
+		private final String property;
+
+		private final ColumnFamilyHandle handle;
+
+		private final RocksDB db;
+
+		private volatile boolean open;
+
+		private long value;
+
+		private RocksDBNativeMetricView(
+			@Nonnull String property,
+			@Nonnull ColumnFamilyHandle handle,
+			@Nonnull RocksDB db
+		) {
+			this.property = property;
+			this.handle = handle;
+			this.db = db;
+			this.open = true;
+		}
+
+		@Override
+		public String getValue() {
+			return Long.toUnsignedString(value);
+		}
+
+		@Override
+		public void update() {
+			if (open) {
```

Comment:
I think this can potentially cause a segfault. A metrics related thread can call `update()` and pass the `if`-check. In the meantime the main thread can call `close()` on the backend, this propagates, the flag is set, the lease is released, the native db can be closed and then the update-caller could continue execution, accessing the released native resource in `db`.

Reply:
One way of doing this correctly could be to aquire a lease before reading and releasing the lease when the read is complete. Using state transitions (open, readlocked, closed) via AtomicRefernces could also work. However, I think there is potential for better approaches if the metrics system supports it, e.g. if there is a threadpool used to run the updates per task, then we could first shutdown the metrics threads and then dispose other resources like the backends. What would be desirable is that the metrics pulling threads are pooled, and when the pool is closed we are sure that there are no more in-flight update calls. This could eliminate the need for `volatile`, `lease` or other synchronnization measures.
@zentol is there a way to shut down the metrics pulling in such a fashion, where we are sure that no more updates are pending or triggered in the future?

Reply:
There exists only a single thread that does the polling, which is responsible for running reporters, so it can only be shutdown when the cluster as a whole terminates. If the `MetricRegistry` is shutdown it is guaranteed that no polling happens anymore.

A nicer way to do this would be to unregister the metrics (by closing the group they were registered on) and enhance the [ViewUpdaterTask](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/metrics/ViewUpdater.java) to immediately react to unregistered metrics.

Reply:
Ok, that is also how I understood they way it works. I guess for proper lifecycle management it would be really good to have a way that cleanly stops (some) metrics from updating.

Reply:
I create a `MetricGroup` per column family so keeping references to each of those would be simple. If we just make `view` closeable and each gauge keeps a lease for its entire lifetime then we can close all the metrics groups in dispose which will unregister all the gauges and close the leases. Thoughts?

Reply:
Well, there is also option 3, use an atomic reference with 3 states and use compareAndSet to guard the critical section in update and set to a terminal state in close. Getting and releasing a lease is currently used for the long lasting operation of checkpointing. This has a similar effect to stop the thread from accessing db after close, but is a bit cheaper.